src/esssans/i_of_q.py

```diff
@@ -177,36 +195,54 @@ def _to_q_bins(q_bins: Union[int, sc.Variable]) -> Dict[str, Union[int, sc.Varia
 def _events_merge_spectra(
     data_q: sc.DataArray,
     q_bins: Union[int, sc.Variable],
+    final_dims: List[str],
     wavelength_bands: Optional[sc.Variable] = None,
 ) -> sc.DataArray:
     """
     Merge spectra of event data
     """
-    q_all_pixels = data_q.bins.concat(set(data_q.dims) - {'Q'})
+    q_all_pixels = data_q.bins.concat(set(data_q.dims) - set(final_dims))
     edges = _to_q_bins(q_bins)
-    if wavelength_bands is not None:
-        edges[wavelength_bands.dim] = wavelength_bands
-    return q_all_pixels.bin(**edges)
+    q_binned = q_all_pixels.bin(**edges)
+    if wavelength_bands is None:
+        return q_binned
+
+    # Note: cannot simply add the wavelength_bands to the edges, because the
+    # wavelength_bands may overlap. So instead, we loop over the bands and bin each
+    # band separately.
+    #
+    # Note: it is more memory efficient to concatenate the bands as we loop, rather
+    # than to keep all bands in a list and then concatenate the results.
+    band_dim = (set(wavelength_bands.dims) - {'wavelength'}).pop()
+    out = None
+    for wav_range in sc.collapse(wavelength_bands, keep='wavelength').values():
+        band = q_binned.bin(wavelength=wav_range).squeeze()
+        if out is None:
+            out = band
+        else:
+            out = sc.concat([out, band], band_dim)
+    return out
```

Comment:
Any good ideas on how we could do this differently?

Reply:
Refactored from the comment above to use a `_process_wavelength_bands` function that performs some checks and some optional reshaping.

Reply:
It would probably be faster if we first apply some binning in wavelength, which will allow for pre-selecting bins. But that is already an optimization, I'd rather first think about avoiding this.

Unless bands are overlapping, can we revert to the old approach?

Which gives me an idea:
1. Flatten and sort all the wavelength edges (potentially overlapping).
2. `bin`
3. Assemble bands by combining the (probably up to 3) pieces for each band.

Reply:
Something like this:
```python
def joint(da: sc.DataArray, bands: sc.Variable) -> sc.DataArray:
    binned = da.bin(y=sc.sort(bands.flatten(to='y'), 'y'))
    # At this point we kind of already have what we need, would be cheapest to just return,
    # if follow up providers can work with the result. Otherwise we need to duplicate events:
    sections = []
    for bounds in sc.collapse(bands, keep='y').values():
        # The extra concat can probably be avoided if we insert some dummy edges for first and last band
        sections.append(binned['y',bounds[0]:bounds[1]].bins.concat('y'))
    out = sc.concat(sections, 'band')
    out.coords['y'] = bands
    return out
```

Reply:
> # The extra concat can probably be avoided if we insert some dummy edges for first and last band

This would only work if we know how many edges to insert, as there is no limit on how wide the bands can be; they could overlap by more than one bin?